From 776ff3ea2f39a0b55e8a01749871334620d5dc71 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:05:03 -0400
Subject: [powerpc] powernv: Return secondary CPUs to firmware on kexec

Message-id: <1410545655-205645-75-git-send-email-dzickus@redhat.com>
Patchwork-id: 94090
O-Subject: [RHEL7 PATCH 074/626] powerpc/powernv: Return secondary CPUs to firmware on kexec
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 13906db670a128864714c30c244b866dce119494
Author: Benjamin Herrenschmidt <benh@kernel.crashing.org>
Date:   Wed Aug 21 13:03:20 2013 +1000

    powerpc/powernv: Return secondary CPUs to firmware on kexec

    With OPAL v3 we can return secondary CPUs to firmware on kexec. This
    allows firmware to do various cleanups making things generally more
    reliable, and will enable the "new" kernel to call OPAL to perform
    some reconfiguration tasks early on that can only be done while
    all the CPUs are in firmware.

    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/include/asm/opal.h b/arch/powerpc/include/asm/opal.h
index 07b55a7..a865b07 100644
--- a/arch/powerpc/include/asm/opal.h
+++ b/arch/powerpc/include/asm/opal.h
@@ -128,6 +128,7 @@ extern int opal_enter_rtas(struct rtas_args *args,
 #define OPAL_XSCOM_WRITE   66
 #define OPAL_LPC_READ    67
 #define OPAL_LPC_WRITE    68
+#define OPAL_RETURN_CPU    69
 
 #ifndef __ASSEMBLY__
 
@@ -711,6 +712,7 @@ int64_t opal_set_system_attention_led(uint8_t led_action);
 int64_t opal_pci_next_error(uint64_t phb_id, uint64_t *first_frozen_pe,
        uint16_t *pci_error_type, uint16_t *severity);
 int64_t opal_pci_poll(uint64_t phb_id);
+int64_t opal_return_cpu(void);
 
 int64_t opal_xscom_read(uint32_t gcid, uint32_t pcb_addr, uint64_t *val);
 int64_t opal_xscom_write(uint32_t gcid, uint32_t pcb_addr, uint64_t val);
diff --git a/arch/powerpc/platforms/powernv/opal-wrappers.S b/arch/powerpc/platforms/powernv/opal-wrappers.S
index c9e67a8..2a03e1e 100644
--- a/arch/powerpc/platforms/powernv/opal-wrappers.S
+++ b/arch/powerpc/platforms/powernv/opal-wrappers.S
@@ -122,3 +122,4 @@ OPAL_CALL(opal_xscom_read,   OPAL_XSCOM_READ);
 OPAL_CALL(opal_xscom_write,   OPAL_XSCOM_WRITE);
 OPAL_CALL(opal_lpc_read,   OPAL_LPC_READ);
 OPAL_CALL(opal_lpc_write,   OPAL_LPC_WRITE);
+OPAL_CALL(opal_return_cpu,   OPAL_RETURN_CPU);
diff --git a/arch/powerpc/platforms/powernv/setup.c b/arch/powerpc/platforms/powernv/setup.c
index 4ddb339..e239dcf 100644
--- a/arch/powerpc/platforms/powernv/setup.c
+++ b/arch/powerpc/platforms/powernv/setup.c
@@ -31,6 +31,7 @@
 #include <asm/xics.h>
 #include <asm/rtas.h>
 #include <asm/opal.h>
+#include <asm/kexec.h>
 
 #include "powernv.h"
 
@@ -153,6 +154,16 @@ static void pnv_shutdown(void)
 static void pnv_kexec_cpu_down(int crash_shutdown, int secondary)
 {
  xics_kexec_teardown_cpu(secondary);
+
+ /* Return secondary CPUs to firmware on OPAL v3 */
+ if (firmware_has_feature(FW_FEATURE_OPALv3) && secondary) {
+  mb();
+  get_paca()->kexec_state = KEXEC_STATE_REAL_MODE;
+  mb();
+
+  /* Return the CPU to OPAL */
+  opal_return_cpu();
+ }
 }
 #endif /* CONFIG_KEXEC */
 
-- 
1.7.1